src/NServiceBus.Core/Pipeline/Incoming/DeserializeMessageConnector.cs

```diff
@@ -93,6 +94,10 @@ LogicalMessage[] Extract(IncomingMessage physicalMessage)
                     log.WarnFormat("Could not determine message type from message header '{0}'. MessageId: {1}", messageTypeIdentifier, physicalMessage.MessageId);
                 }
             }
+            else if (!allowContentTypeInference)
```

Comment:
Could we invert the `if`? Then not have an else to not further indent the code?

Reply:
do you mean something like inverting the field so we can `else if(disableContentTypeInference)` ?

Reply:
talking to Ramon, the suggestion is `if (!physicalMessage.Headers.TryGetValue(Headers.EnclosedMessageTypes, out var messageTypeIdentifier) && !allowContentTypeInference ){ throw ... }` instead.

👍

Reply:
the inversion actually doesn't work as we have to deal with 3 cases that can't be represented nicely when inversing:

* headers are present -> fetch message type from headers
* headers are not present && type inference is allowed -> skip message detection and go straight to serializer
* headers are not present && type inference is disabled -> throw error